### ISSUE TITLE
No support for Method chaining

### ISSUE THREAD

**Issue by Jictyvoo** [vlang/v#2968]
**V version:** https://github.com/vlang/v/tree/47f5e1178ff308b87267ea07856f2a6574cfd84b
**OS:** Ubuntu 18

**What did you do?**
I've created a struct with method chaining and it gives a C error
```v
module main

import strings

struct MethodChaining {
	history strings.Builder
}

fn (m mut MethodChaining) home() MethodChaining { m.history.write("home ") return m }

fn (m mut MethodChaining) car() MethodChaining { m.history.write("car ") return m }

fn (m mut MethodChaining) work() MethodChaining { m.history.write("work ") return m }

fn (m mut MethodChaining) gym() MethodChaining { m.history.write("gym ") return m }

fn (m mut MethodChaining) end() MethodChaining { m.history.write("end ") return m }

fn (m mut MethodChaining) str() string { return m.history.str() }

fn main() {
	mut m := MethodChaining{}
	println(m.home().car().work().car().gym().car().home().end().str())
}
```

If import http have another error, just only importing http module

I think that's some error in pointer handling
**What did you expect to see?**
To work
 
**What did you see instead?**
C error

**Comment by Jictyvoo** [comment]
Another error is in execution time, when write main function like:
```v
fn main() {
	mut m := MethodChaining{}
	m = m.home()
	m = m.car()
	m = m.work()
	m = m.car()
	m = m.gym()
	m = m.car()
	m = m.home()
	m = m.end()
	println(m.str())
}
```
it will compile, but when execute, receive this error
```shell
V panic: tos(): nil string
                                               | 0x7f23fc6037ef | ./chain_method_error(+0x37ef)
                                               | 0x7f23fc606015 | ./chain_method_error(+0x6015)
                                               | 0x7f23fc60a174 | ./chain_method_error(+0xa174)
                                               | 0x7f23fc6196a3 | ./chain_method_error(+0x196a3)
                                               | 0x7f23fc619881 | ./chain_method_error(+0x19881)
                                               | 0x7f23fc619cff | ./chain_method_error(+0x19cff)
                                               | 0x7f23fb4a1b97 | /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)
                                               | 0x7f23fc6021ea | ./chain_method_error(+0x21ea)
```

**Comment by lemoncmd** [comment]
This works fine
```
fn (m mut MethodChaining) home() &MethodChaining { m.history.write("home ") return m }
```

In the first example, `home()` returns a copy of `m` value (not reference) due to the quick fix for #230 . The returned value can't be referred while chained method requires its reference. This causes C error.

**Comment by Jictyvoo** [comment]
> This works fine
> 
> ```
> fn (m mut MethodChaining) home() &MethodChaining { m.history.write("home ") return m }
> ```
> 
> In the first example, `home()` returns a copy of `m` value (not reference) due to the quick fix for #230 . The returned value can't be referred while chained method requires its reference. This causes C error.

Ok, I changed all `return m` by `return *m`, now second example compile, but the first example with
```v
println(m.home().car().work().car().gym().car().home().end().str())
```
doesn't compile

**Comment by penguindark** [comment]
Solved.
example:
```v
struct A1 {
pub mut:
 a int
}

fn (a A1) m1() A1 {
	return A1{a:a.a + 1}
}

fn (a A1) m2() A1 {
	return A1{a:a.a + 2}
}

fn main() {
	mut v := A1{}
	println(v.m1().m2())
}
```
